src/services/organizeImports.ts

```diff
@@ -0,0 +1,291 @@
+            const replaceOptions = {
+                useNonAdjustedStartPosition: false,
+                useNonAdjustedEndPosition: false,
+                suffix: getNewLineOrDefaultFromHost(host, formatContext.options),
+             };
+            changeTracker.replaceNodeWithNodes(sourceFile, oldImportDecls[0], newImportDecls, replaceOptions);
+        }
+
+        const changes = changeTracker.getChanges();
+        return changes;
+    }
+
+    function removeUnusedImports(oldImports: ReadonlyArray<ImportDeclaration>) {
+        return oldImports; // TODO (https://github.com/Microsoft/TypeScript/issues/10020)
+    }
+
+    /* @internal */ // Internal for testing
+    export function sortImports(oldImports: ReadonlyArray<ImportDeclaration>) {
+        if (oldImports.length < 2) {
+            return oldImports;
+        }
+
+        // NB: declaration order determines sort order
+        const enum ModuleNameKind {
+            NonRelative,
+            Relative,
+            Invalid,
+        }
+
+        const importRecords = oldImports.map(createImportRecord);
+
+        const sortedRecords = stableSort(importRecords, (import1, import2) => {
+            const { name: name1, kind: kind1 } = import1;
+            const { name: name2, kind: kind2 } = import2;
+
+            if (kind1 !== kind2) {
+                return kind1 < kind2
+                    ? Comparison.LessThan
+                    : Comparison.GreaterThan;
+            }
+
+            // Note that we're using simple equality, retaining case-sensitivity.
+            if (name1 !== name2) {
+                return name1 < name2
+                    ? Comparison.LessThan
+                    : Comparison.GreaterThan;
+            }
+
+            return Comparison.EqualTo;
+        });
+
+        return sortedRecords.map(r => r.importDeclaration);
+
+        function createImportRecord(importDeclaration: ImportDeclaration) {
+            const specifier = importDeclaration.moduleSpecifier;
+            const name = getExternalModuleName(specifier);
+            if (name) {
+                const isRelative = isExternalModuleNameRelative(name);
+                return { importDeclaration, name, kind: isRelative ? ModuleNameKind.Relative : ModuleNameKind.NonRelative };
+            }
+
+            return { importDeclaration, name: specifier.getText(), kind: ModuleNameKind.Invalid };
+        }
+    }
+
+    function getExternalModuleName(specifier: Expression) {
+        return isStringLiteral(specifier) || isNoSubstitutionTemplateLiteral(specifier)
+            ? specifier.text
+            : undefined;
+    }
+
+    /**
+     * @param sortedImports a non-empty list of ImportDeclarations, sorted by module name.
+     */
+    function groupSortedImports(sortedImports: ReadonlyArray<ImportDeclaration>): ReadonlyArray<ReadonlyArray<ImportDeclaration>> {
+        Debug.assert(length(sortedImports) > 0);

```

Comment:
Maybe better:
```ts
function groupSortedImports(sortedImports: ReadonlyArray<ImportDeclaration>): ReadonlyArray<ReadonlyArray<ImportDeclaration>> {
    return groupBy(sortedImports, (a, b) => getExternalModuleName(a.moduleSpecifier) === getExternalModuleName(b.moduleSpecifier));
}

function groupBy<T>(values: ReadonlyArray<T>, areSameGroup: (a: T, b: T) => boolean): ReadonlyArray<ReadonlyArray<T>> {
    Debug.assert(values.length !== 0);
    const groups: T[][] = [];
    let group: T[] = [];
    for (const value of values) {
        if (group.length && !areSameGroup(value, group[0])) {
            groups.push(group);
            group = [];
        }
        group.push(value);
    }
    if (group.length) {
        groups.push(group);
    }
    return groups;
}
```

Reply:
Just to confirm, your improvement was pulling out a generic helper?  The actual behavior is the same?

Reply:
Yes, assuming invalid imports are filtered out (so `getExternalModuleName`) returns a defined result).

Reply:
Just noticed we don't need `Debug.assert(values.length !== 0);` in the "maybe better" version.

Reply:
Note that this isn't true group-by - it requires that the input be sorted.  Is it still worth pulling out a generic helper for specialized functionality?

Reply:
Actually, this doesn't *just* rely on it being sorted, but on it being sorted by the same mechanism as `areSameGroup`. Maybe it would be better to group first (not relying on sorting) and then sort the groups?

Reply:
Yes, that's why it's a private helper function.  At the moment, there doesn't seem to be a need for something more general, does there?

Reply:
As you requested, we are now grouping before performing the other operations.